apis/Google.Cloud.Metadata.V1/Google.Cloud.Metadata.V1.IntegrationTests/Emulator/metadata_server.py

```diff
@@ -0,0 +1,353 @@
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from collections import OrderedDict
+from flask import Flask, request
+from werkzeug.wrappers import Request
+import common, errno, getopt, json, logging, os, socket, sys, time
+
+numeric_project_id = 12345;
+numeric_project_id_token = "<NUMERIC_PROJECT_ID>";
+project_id = 'fake-project';
+project_id_token = "<PROJECT_ID>";
+
+app = Flask(__name__)
+
+def excludeKeyInRecursiveGet(key):
+  return key == 'token'
+
+def formatResponse(data, status_code, content_type):
+  return data, status_code, {common.content_type: content_type}
+
+def formatOrderedDictContents(data):
+  def getKeyDescription(key):
+    if not excludeKeyInRecursiveGet(key):
+      if type(data[key]) is OrderedDict or isinstance(data[key], IndexedList):
+        return key + '/\n'
+
+    return key + '\n'
+
+  return ''.join([getKeyDescription(key) for key in data])
+
+class HeaderCheckMiddleware(object):
+  def __init__(self, app):
+     self.app = app
+
+  def __call__(self, environ, start_response):
+    req = Request(environ, shallow=True)
+    if (req.headers.get(common.metadata_flavor) != common.metadata_flavor_google):
+      start_response('403 Forbidden',
+                     [(common.content_type, common.content_type_html),
+                      (common.metadata_flavor, common.metadata_flavor_google),
+                      (common.server, common.server_value)])
+      return ['Missing Metadata-Flavor:Google header.']
+    return self.app(environ, start_response)
+
+# Collection whose content is accessed via an index value in the relative URL.
+class IndexedList:
+  def __init__(self, list):
+    self._list = list
+
+  def __iter__(self):
+    return self._list.__iter__()
+
+  def __len__(self):
+    return self._list.__len__()
+
+  def __delitem__(self, key):
+    self._list.__delitem__(key)
+
+  def __getitem__(self, key):
+    return self._list.__getitem__(key)
+  
+  def __setitem__(self, key, value):
+    self._list.__setitem__(key, value)
+
+  def formatContents(self):
+    return ''.join([unicode(i) + '/\n' for i in range(len(self._list))])
+
+class MetadataJSONEncoder(json.JSONEncoder):
+  def fixupOrderedDict(self, d):
+    result = OrderedDict()
+    for key in d:
+      if excludeKeyInRecursiveGet(key):
+        continue
+
+      if key.endswith('@developer.gserviceaccount.com'):
+        newKey = key
+      else:
+        newKey = common.convertToCamelCase(key)
+
+      if (d[key] == numeric_project_id_token):
+        result[newKey] = numeric_project_id
+      else:
+        result[newKey] = self.default(d[key])
+
+    return result
+
+  def encode(self, o):
+    if isinstance(o, OrderedDict):
```

Comment:
why not just have a method like o2 = ensureOrderedDict(o), that way instead of always checking if it is an OrderedDict, we just always call it, and ensureOrderedDict can be responsible for checking types etc... I generally like to contain type checks etc where possible